### ISSUE TITLE
TabController calls listeners twice when tab changes

### ISSUE THREAD

**Issue by alibitek** [flutter/flutter#13848]
While playing with [TabController class](https://docs.flutter.io/flutter/material/TabController-class.html) I noticed that when **tapping** between tabs in the AppBar there are two renderings (i.e. the listener function is called twice, not once, thus generating two rebuilds), but when **swiping** there is only one rebuild.

I've made a [testcase application](https://github.com/alibitek/tab_controler_double_notify) that showcases the problem.

The code of the test widget is [here](https://github.com/alibitek/tab_controler_double_notify/blob/master/lib/main.dart#L18)

The listener is added as follows:
```dart
@override
  void initState() {
    super.initState();

   //  ...
  _tabController = new TabController(vsync: this, initialIndex: 0, length: MyWidgetTabs.values.length);
  _tabController.addListener(_handleTabSelection);
  //  ...
}

 void _handleTabSelection() {
    setState(() {
      _selectedTab = MyWidgetTabs.values[_tabController.index];
      print(
          "Changed tab to: ${_selectedTab.toString().split('.').last} , index: ${_tabController.index}");
    });
  }
```

The problematic behaviour comes from the [TabController's _changeIndex method](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/tab_controller.dart#L109):
```dart
  void _changeIndex(int value, { Duration duration, Curve curve }) {
    assert(value != null);
    assert(value >= 0 && (value < length || length == 0));
    assert(duration == null ? curve == null : true);
    assert(_indexIsChangingCount >= 0);
    if (value == _index || length < 2)
      return;
    _previousIndex = index;
    _index = value;
    if (duration != null) {
      _indexIsChangingCount += 1;
      notifyListeners(); // Because the value of indexIsChanging may have changed.
      _animationController
        .animateTo(_index.toDouble(), duration: duration, curve: curve)
        .whenCompleteOrCancel(() {
          _indexIsChangingCount -= 1;
          notifyListeners();
        });
    } else {
      _indexIsChangingCount += 1;
      _animationController.value = _index.toDouble();
      _indexIsChangingCount -= 1;
      notifyListeners();
    }
}
```
Changes introduced in pull requests: #10016 and  #8741

When **swiping** it goes to else code block, and when **tapping** one of the TabBar's Tab's in the AppBar it goes inside the if, which calls `notifyListeners()` before the animation starts and after the animation ends.

## Output logs

Swipe from tab 1 to tab2 logs:
```
I/flutter (20727): Changed tab to: tab2 , index: 1
I/flutter (20727): Build - main
I/flutter (20727): Build tab - MyWidgetTabs.tab1
I/flutter (20727): Build tab - MyWidgetTabs.tab2
```

Tap tab2 from tab1 logs:
```
I/flutter (20727): Changed tab to: tab2 , index: 1
I/flutter (20727): Build - main
I/flutter (20727): Build tab - MyWidgetTabs.tab1
I/flutter (20727): Build tab - MyWidgetTabs.tab2
I/flutter (20727): Changed tab to: tab2 , index: 1
I/flutter (20727): Build - main
I/flutter (20727): Build tab - MyWidgetTabs.tab1
I/flutter (20727): Build tab - MyWidgetTabs.tab2
```

The correct behaviour when tapping the AppBar's tab2 , when current tab is tab1 should be:
```
I/flutter ( 1361): Changed tab to: tab2 , index: 1
I/flutter ( 1361): Build - main
I/flutter ( 1361): Build tab - MyWidgetTabs.tab1
I/flutter ( 1361): Build tab - MyWidgetTabs.tab2
```

**Comment by PixelToast** [comment]
Notifying listeners is intended behavior since it's updating `_indexIsChangingCount`, this PR may fix the rebuild issue but it does introduce a bug.

**Comment by Hixie** [comment]
As far as I can tell this is operating as expected. When you swipe, the tab controller doesn't actually change its state until the end of the drag. When you tap, we do an animation which causes the state to change twice, once at the start and once at the end. Thus we have to notify twice.

Generally speaking, build methods should be very fast and should be able to handle being called more times than strictly necessary. If nothing has changed, the effect on performance will be minimal (e.g. we won't do a layout or repaint).

That no tests failed when you remove that line of code is not a good sign for our test coverage. :-)

cc @HansMuller

**Comment by alibitek** [comment]
In the else case, `notifyListeners()` is not called after every _indexIsChangingCount change, even though technically the objects' data has changed.

The `indexIsChanging` property can be used the same way to check if there is an animated change of index in progress, as it is backed by `_indexIsChangingCount` :
```
  if (!_tabController.indexIsChanging)
      _tabController.animateTo((_tabController.index + delta).clamp(0, MyWidgetTabs.values.length - 1));
```

I see that the TabBar adds a listener to the TabController in [_handleTabControllerTick](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/tabs.dart#L714):
```
void _handleTabControllerTick() {
    if (_controller.index != _currentIndex) {
      _currentIndex = _controller.index;
      if (widget.isScrollable)
        _scrollToCurrentIndex();
    }
    setState(() {
      // Rebuild the tabs after a (potentially animated) index change
      // has completed.
    });
}
```

Whenever the TabController notifies the TabBar it rebuilds the tabs.

Q1: I think the contention is whether TabController should call `notifyListeners()` before or after the animation or both?

Made some tests by commenting the `notifyListeners();` call before and after the animation.

With `notifyListeners()` *before and after animation**:
```
I/flutter (13602): Changed tab to: tab2 , index: 1
I/flutter (13602): _handleTabControllerTick - _controller.index: 1 , _currentIndex: 0
I/flutter (13602): Build - main
I/flutter (13602): Build tab - MyWidgetTabs.tab1
I/flutter (13602): Build tab - MyWidgetTabs.tab2
I/flutter (13602): Changed tab to: tab2 , index: 1
I/flutter (13602): _handleTabControllerTick - _controller.index: 1 , _currentIndex: 1
I/flutter (13602): animation finished
I/flutter (13602): Build - main
I/flutter (13602): Build tab - MyWidgetTabs.tab1
I/flutter (13602): Build tab - MyWidgetTabs.tab2
```

With `notifyListeners()` **only before animation**:
```
I/flutter (13602): Changed tab to: tab2 , index: 1
I/flutter (13602): _handleTabControllerTick - _controller.index: 1 , _currentIndex: 0
I/flutter (13602): Build - main
I/flutter (13602): Build tab - MyWidgetTabs.tab1
I/flutter (13602): Build tab - MyWidgetTabs.tab2
I/flutter (13602): animation finished
```

With `notifyListeners()` **only after animation**:
```
01-02 23:58:13.853 13602 13673 I flutter : ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
01-02 23:58:13.854 13602 13673 I flutter : The following assertion was thrown building _TabStyle(animation: _DragAnimation#88ef1(▶), dirty,
01-02 23:58:13.854 13602 13673 I flutter : state: _AnimatedState#4be3a):
01-02 23:58:13.854 13602 13673 I flutter : 'package:flutter/src/material/tabs.dart': Failed assertion: line 404 pos 12:
01-02 23:58:13.854 13602 13673 I flutter : '!controller.indexIsChanging': is not true.
01-02 23:58:13.895 13602 13673 I flutter : 
01-02 23:58:13.896 13602 13673 I flutter : Either the assertion indicates an error in the framework itself, or we should provide substantially
01-02 23:58:13.896 13602 13673 I flutter : more information in this error message to help you determine and fix the underlying cause.
01-02 23:58:13.896 13602 13673 I flutter : In either case, please report this assertion by filing a bug on GitHub:
01-02 23:58:13.896 13602 13673 I flutter :   https://github.com/flutter/flutter/issues/new
01-02 23:58:13.896 13602 13673 I flutter : 
01-02 23:58:13.896 13602 13673 I flutter : When the exception was thrown, this was the stack:
01-02 23:58:13.897 13602 13673 I flutter : #2      _DragAnimation.value (package:flutter/src/material/tabs.dart:404:12)
01-02 23:58:13.897 13602 13673 I flutter : #3      _TabStyle.build (package:flutter/src/material/tabs.dart:142:62)
01-02 23:58:13.897 13602 13673 I flutter : #4      _AnimatedState.build (package:flutter/src/widgets/transitions.dart:96:48)
01-02 23:58:13.898 13602 13673 I flutter : #5      StatefulElement.build (package:flutter/src/widgets/framework.dart:3690:27)
01-02 23:58:13.898 13602 13673 I flutter : #6      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3602:15)
01-02 23:58:13.898 13602 13673 I flutter : #7      Element.rebuild (package:flutter/src/widgets/framework.dart:3463:5)
01-02 23:58:13.898 13602 13673 I flutter : #8      BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2210:33)
01-02 23:58:13.898 13602 13673 I flutter : #9      BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:612:20)
01-02 23:58:13.898 13602 13673 I flutter : #10     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:208:5)
01-02 23:58:13.898 13602 13673 I flutter : #11     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:921:15)
01-02 23:58:13.898 13602 13673 I flutter : #12     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:861:9)
01-02 23:58:13.898 13602 13673 I flutter : #13     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:773:5)
01-02 23:58:13.898 13602 13673 I flutter : #14     _invoke (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:120)
01-02 23:58:13.899 13602 13673 I flutter : #15     _drawFrame (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:109)
01-02 23:58:13.899 13602 13673 I flutter : (elided 2 frames from class _AssertionError)
01-02 23:58:13.899 13602 13673 I flutter : ════════════════════════════════════════════════════════════════════════════════════════════════════
01-02 23:58:13.907 13602 13673 I flutter : Another exception was thrown: 'package:flutter/src/material/tabs.dart': Failed assertion: line 404 pos 12: '!controller.indexIsChanging': is not true.
01-02 23:58:13.918 13602 13673 I flutter : Another exception was thrown: A RenderFlex overflowed by 99872 pixels on the bottom.
01-02 23:58:13.943 13602 13673 I flutter : Another exception was thrown: 'package:flutter/src/material/tabs.dart': Failed assertion: line 404 pos 12: '!controller.indexIsChanging': is not true.
01-02 23:58:14.142 13602 13673 I flutter : Changed tab to: tab1 , index: 0
01-02 23:58:14.142 13602 13673 I flutter : _handleTabControllerTick - _controller.index: 0 , _currentIndex: 1
01-02 23:58:14.142 13602 13673 I flutter : animation finished
01-02 23:58:14.144 13602 13673 I flutter : Build - main
01-02 23:58:14.144 13602 13673 I flutter : Build tab - MyWidgetTabs.tab1
01-02 23:58:14.144 13602 13673 I flutter : Build tab - MyWidgetTabs.tab2
```

Q2: With my change, after the animation, the widgets are no longer rebuilt. Can this be an issue?

Q3: To be notified **only once** by the TabController when the tab has been changed, would it make sense to add this functionality to the class?
For example:
- by adding a callback function to the TabController constructor?
- a TabChanged class as a Stream?
- a TabChanged class with ChangeNotifier?
- any other suggestion?

**Comment by Hixie** [comment]
> In the else case, `notifyListeners()` is not called after every _indexIsChangingCount change, even though technically the objects' data has changed.

Right, because `_indexIsChangingCount` isn't exposed, only `indexIsChanging` is exposed, and in the else case, that doesn't change.

> Q1: I think the contention is whether TabController should call `notifyListeners()` before or after the animation or both?

It should call it whenever its properties change. (Note that it _doesn't_ get called when the _animation_ is changing, because that's a separate property that also separately implements `Listenable`.)

> Q2: With my change, after the animation finishes, the widgets are no longer rebuilt. They're only built before the animation. Can this be an issue?

The issue is just that if a client of the TabController is expecting to be notified whenever `indexIsChanging` changes, they won't be notified in all cases any more, and so could get out of sync.

> Q3: To be notified **only once** by the TabController when the tab has been changed

It shouldn't be a goal to be only notified once. You could be notified arbitrarily at any time, and your build methods should be ready for this. For example, it's common for build methods to be called whenever the screen rotates, or if the user changes locales, or if the theme changes, etc.

**Comment by volgin** [comment]
> It shouldn't be a goal to be only notified once. You could be notified arbitrarily at any time, and your build methods should be ready for this. 

When  a tab change requires an async call, we have a problem. If the second notification fires before async returns, the async method will be called again. This is probably a typical use case, e.g. lazy loading data into tabs.

Multiple notifications is not an issue *if* they distinguish themselves, as in "press down" / "press up".

**Comment by Hixie** [comment]
Async work should be set off during initState, not during build. initState will be called only once.

**Comment by SSA111** [comment]
As volgin denotes would it be possible to add some kind of context to the notification instead? I have experienced cases where the tabbar seems to entirely skip the animation due to a rebuild being done while the animation is in process. As the listener is being called multiple times, it causes multiple rebuilds if you have state changing logic in your listener method that relates to the tabbar itself, for example  setting the index of the tabbar (I should be able to code up a simple example, if wanted). 
Would it be an idea to go down the path of how NSNotification is structured in iOS, that includes a key identifing the specific notification sent? Would that cause any issues? Nevertheless I would assume this approach only adds an increased flexibility for the developer, while taking nothing away from its current form.

**Comment by kevdev03** [comment]
hello @alibitek, have you managed to overcome this issue? I'm currently experiencing it now and I'm at my wit's end. SOS. thank you

**Comment by pedromorgan** [comment]
So @ferdigokdemir code above with `_tabController.indexIsChanging` works..

It is strange behaviour that its double triggering imho as a newbie.

**Comment by jeysenlnoy1993** [comment]
Hi Folks! if above answer don't work try this one :)
*inside the listener

       if(tabController.indexIsChanging) {
          print("tab is animating. from active (getting the index) to inactive(getting the index) ");
       }else {
           //tab is finished animating you get the current index
          //here you can get your index or run some method once.
          print(tabController.index);
       }

hope this will helps :D

**Comment by BaranMichal25** [comment]
If someone would be interested in getting more immediate updates to achieve more smooth effects, you can also listen for animation changes and set state accordingly.
Here's an example for 2 tabs scenario:
```dart
class HomeScreen extends StatefulWidget {
  @override
  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen>
    with SingleTickerProviderStateMixin {
  int index = 0;
  TabController _tabController;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(vsync: this, length: 2);
    _tabController.animation.addListener(onTabChanged);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: ..,
      body: ...,
      floatingActionButton: index == 0
          ? FloatingActionButton(
              child: Icon(Icons.add),
              onPressed: () {},
            )
          : null,
    );
  }

  void onTabChanged() {
    final aniValue = _tabController.animation.value;
    if (aniValue > 0.5 && index != 1) {
      setState(() {
        index = 1;
      });
    } else if (aniValue <= 0.5 && index != 0) {
      setState(() {
        index = 0;
      });
    }
  }
}
```
This will trigger `setState()` as soon as user swipes more than 50% of area.

**Comment by aleksandar-radivojevic** [comment]
@BaranMichal25 Can you please show me how to do it when I have 7 tabs? I just cant figure it out. Thanks bro 🚀

**Comment by ifredom** [comment]
This bug monitoring will trigger twice. When can it be fixed？

**Comment by abhipoojary** [comment]
this issue is not sorted, may i know when can it be fixed?

**Comment by Nandhini-Ravichandran** [comment]
This bug not resolved, How to fix this?

**Comment by PixelToast** [comment]
@Nandhini-Ravichandran @abhipoojary
TabController notifying listeners twice is intended behavior, see the comment at https://github.com/flutter/flutter/issues/13848#issuecomment-354887932

If you have a proposed change feel free to make a [feature request](https://github.com/flutter/flutter/issues/new?assignees=&labels=&template=3_feature_request.md&title=) so the team can look at it.

**Comment by pedromorgan** [comment]
```if (!_tabController.indexIsChanging)```
